Convert more classes to record classes

shubhamvishu opened the pull request:
### Description

- This PR addresses #13207 to convert more classes on `main` branch to record classes on main (Lucene 10 only).
- It moves a lot of data classes(120 to be precise) to use record classes that were flagged suitable in my IDE and seemed good candidate in general.
- I left a couple of them like `TotalHits` and `SynonymMap` as the PR is already very big and including those were leading to a lot more changes. Maybe we could do those as a separate PR.

&nbsp;
Raising this PR since all the tests are passing(`./gradlew test`) but `renderJavadoc` task is complaining about missing java docs on some record classes(converted in this PR) which I see has the javadocs already eg: `TermStats`, `ReaderSlice`. I'm not sure why its flagging those incorrectly or if maybe I'm missing something.

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

shubhamvishu committed:
Convert more classes to record classes

shubhamvishu commented on lucene/core/src/java/org/apache/lucene/codecs/TermStats.java:
```diff
@@ -16,24 +16,10 @@
-    this.totalTermFreq = totalTermFreq;
-  }
-}
+public record TermStats(int docFreq, long totalTermFreq) {}
```
The `renderJavaDoc` fails for this even though java docs are present. If I see earlier PR #13205 that moved some classes I don't see it failing for such classes. If this is not expected, then maybe we need to fix something in `MissingDoclet.java`?(not sure).

cc - @uschindler

uschindler replied:
What's exactly he message?

uschindler replied:
I could only imagine that MissingDoclet can't correctly handle inner classes!? But on the other hand it works fine with enums and it also operates recursively.

I have no time now to check all this out, so the error message and/or stack trace would be good.

uschindler replied:
oh it is a top-level class. no idea whats wrong!

shubhamvishu replied:
Here is what the failure message looks like :
 
```
> Task :lucene:core:renderSiteJavadoc
/home/runner/work/lucene/lucene/lucene/core/src/java/org/apache/lucene/codecs/TermStats.java:25: error: TermStats.<init> (constructor): javadocs are missing
public record TermStats(int docFreq, long totalTermFreq) {}
       ^
error: TermStats.docFreq (method): javadocs are missing
error: TermStats.totalTermFreq (method): javadocs are missing
/home/runner/work/lucene/lucene/lucene/core/src/java/org/apache/lucene/index/ReaderSlice.java:27: error: ReaderSlice.<init> (constructor): javadocs are missing
public record ReaderSlice(int start, int length, int readerIndex) {
       ^
error: ReaderSlice.start (method): javadocs are missing
error: ReaderSlice.length (method): javadocs are missing
error: ReaderSlice.readerIndex (method): javadocs are missing
```

uschindler replied:
The fix should be a separate PR, as this is too large. Maybe we do some small PR first with only the TermStats class and make it pass the tester. Would this be OK to you? I will help to fix the linter.

uschindler replied:
I think the fix needs to be done here: https://github.com/apache/lucene/blob/9af3ef8952597ead37a5a9ce5373ed3bbd40238e/dev-tools/missing-doclet/src/main/java/org/apache/lucene/missingdoclet/MissingDoclet.java#L260

The problem is: how to know which of the ctors and methods are not synthetic?

uschindler replied:
It's complicated: we need to find all accessor methods and automatically generated ctors and exclude them. You can get the accessor method from the record component. Maybe we should build a set of them and exclude them from the checks: https://docs.oracle.com/en/java/javase/21/docs/api/java.compiler/javax/lang/model/element/RecordComponentElement.html

Let's make a separate issue and remove all problematic records from this PR.

uschindler replied:
Maybe as a quick fix we can exclude analysis of record ctors and methods until we have a fix for that. Let's rename that method to `isSyntheticMethod` (remove "enum" from name):
https://github.com/apache/lucene/blob/9af3ef8952597ead37a5a9ce5373ed3bbd40238e/dev-tools/missing-doclet/src/main/java/org/apache/lucene/missingdoclet/MissingDoclet.java#L274-L282

and modify it:

```java
  private boolean isSyntheticMethod(Element element) {
    if (element.getEnclosingElement().getKind() == ElementKind.RECORD) {
      // TODO: fix this to exclude autogenerated accessors and ctors of records, which is hard
      return true;
    }
    String simpleName = element.getSimpleName().toString();
    if (simpleName.equals("values") || simpleName.equals("valueOf")) {
      if (element.getEnclosingElement().getKind() == ElementKind.ENUM) {
        return true;
      }
    }
    return false;
  }
```

uschindler replied:
I think to fix this we have to check for "mandated" elements: https://docs.oracle.com/en/java/javase/21/docs/api/java.compiler/javax/lang/model/util/Elements.Origin.html#MANDATED

Not sure how to do this, so let's do this in a separate issue.

shubhamvishu replied:
Thank you so much for looking into this @uschindler! I like the idea of having a quick fix and exclude the records ctor and methods for now. Also, I even tried to somehow get rid of the earlier issues with records `TermStats` and `ReaderSlice` but explicitly declaring the ctor and methods but it then started complaining about some other public records we added in this PR from the list of packages you shared above where we always expect javadocs. Disabling the check for records till its fixed in a separate issue makes sense to me.💯

shubhamvishu replied:
I opened #13329 with above quickfix change to temporarily disable the javadoc checks and confirmed the tasks passes for this PR with that change.

uschindler replied:
I fixed the issue and added support for records and also its components (as `@param`) in the missing doclet linter: #13332

Once this is merged we can update this PR.

uschindler commented:
Hi, you can merge main into your branch and hopefully checks pass. For some packages like the "codecs" one the new code requires you to also document all record components with `@param`.

uschindler commented on lucene/core/src/java/org/apache/lucene/util/TermAndVector.java:
```diff
@@ -24,23 +24,7 @@
-  public float[] getVector() {
-    return this.vector;
-  }
+public record TermAndVector(BytesRef term, float[] vector) {
```
this class has the problem that it supplies a method to normalize the inner vector. This is against immutability of records. It is possible to do this, but feels bad.

shubhamvishu replied:
Makes sense! I changed it back to a class in the new revision

uschindler commented on lucene/core/src/java/org/apache/lucene/search/CollectionStatistics.java:
```diff
@@ -116,7 +105,8 @@ public CollectionStatistics(
    *
    * @return field's name, not {@code null}
    */
-  public final String field() {
```
we can remove all accessor methods and add the documentation into `@param` on the record declaration.

shubhamvishu replied:
Done! I also changed some more records I could find to be as leaner as possible.

shubhamvishu committed:
Merge branch 'main' into use-record-classes

shubhamvishu committed:
Addressed comments and got all checks passed

shubhamvishu committed:
Made records leaner

shubhamvishu commented on lucene/core/src/java/org/apache/lucene/util/packed/PackedInts.java:
```diff
@@ -171,14 +171,7 @@ public final float overheadRatio(int bitsPerValue) {
+  public record FormatAndBits(Format format, int bitsPerValue) {
 
     @Override
     public String toString() {
```
Removed this now

shubhamvishu commented:
I addressed your comments in the new revision and all the checks are passing now after your fix. Thanks!

uschindler commented:
Thanks, looks fine. I have no time to do another closer review, please give me some time to proceed.

uschindler commented on lucene/codecs/src/java/org/apache/lucene/codecs/blocktreeords/OrdsBlockTreeTermsWriter.java:
```diff
@@ -139,18 +139,17 @@ public final class OrdsBlockTreeTermsWriter extends FieldsConsumer {
-    public final BytesRef maxTerm;
-
-    public FieldMetaData(
+  private record FieldMetaData(
```
make this the default ctor (see above and remove the field assignments), just keep asserts or other checks

shubhamvishu replied:
addressed in the new revision

uschindler commented on lucene/core/src/java/org/apache/lucene/document/NearestNeighbor.java:
```diff
@@ -34,20 +34,16 @@
+  record Cell(
+      PointTree index, int readerIndex, byte[] minPacked, byte[] maxPacked, double distanceSortKey)
+      implements Comparable<Cell> {
+    Cell(
```
remove that ctor and replace by default ctor and only copy minPacked and maxPacked (see above).

shubhamvishu replied:
done

uschindler reviewed (changes requested):
I went through the first few files and have some changes:
- there were some hashCodes revoed but equals survied, please remove both
- some toString may be obsolete, too (unless the exact formatting is important)
- some ctors with exact same parameers like default can be rewritten to use the default syntax and remove the field assignments

I will proceed later on the weekend with other classes.

uschindler commented:
Hi @shubhamvishu,
can you check my comments? The changes here look fine, please fix the remaining issues.
Uwe

shubhamvishu commented:
Hi @uschindler , sorry I was on vacation until last week, so this PR stalled. I'll take a look at the comments today or tomorrow.

shubhamvishu committed:
Merge branch 'main' into use-record-classes

shubhamvishu committed:
Retain LabelAndValue as class after conflict from #13414

shubhamvishu committed:
Remove extra equals,hashCode etc methods and use default constructor syntax

shubhamvishu committed:
Some more changes

shubhamvishu commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:
```diff
@@ -447,14 +446,8 @@ private static int commonCharacterPositionScore(String s1, String s2) {
+      implements Comparable<Weighted<T>> {
 
     @Override
     public boolean equals(Object o) {
```
Removed

shubhamvishu commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/TermAndBoost.java:
```diff
@@ -18,14 +18,13 @@
+ */
+public record TermAndBoost(BytesRef term, float boost) {
   /** Creates a new TermAndBoost */
   public TermAndBoost(BytesRef term, float boost) {
```
Done!.....I changed others such occurrences that I could find as well

shubhamvishu commented:
Hi @uschindler , thanks for the review. I took another pass at the changes and pushed some commits addressing your comments. Please have a look when you get a chance. Thanks!

shubhamvishu committed:
Merge branch 'main' into use-record-classes

shubhamvishu commented:
Hi @uschindler, Could you please review the current code changes once you get some time and if it looks good maybe we can move this forward?

stefanvodita commented:
Added the blocker label to draw attention to this PR for the Lucene 10 release.

shubhamvishu committed:
Merge branch 'main' into use-record-classes

shubhamvishu commented:
Hi @uschindler @mikemccand @jpountz , I fixed the recent merge conflicts and looking for reviews as 10.0 release is blocked on this change. This PR has been attracting merge conflicts frequently since its touching 200+ files, so hoping we could close this sooner. Let me know if there is any concerns here(I'm up for quick followups). Thanks!

jpountz commented on lucene/core/src/java/org/apache/lucene/search/CollectionStatistics.java:
```diff
@@ -17,8 +17,9 @@
-import org.apache.lucene.index.Terms; // javadocs
+
+// javadocs
+// javadocs
```
you can remove these 2 comments

jpountz commented on lucene/misc/src/java/org/apache/lucene/misc/util/fst/UpToTwoPositiveIntOutputs.java:
```diff
@@ -42,36 +42,11 @@
-    @Override
-    public String toString() {
-      return "TwoLongs:" + first + "," + second;
-    }
```
we'll get a different toString() but I think it's fine in this case

shubhamvishu replied:
Yes, its fine in this case.

jpountz commented on lucene/monitor/src/java/org/apache/lucene/monitor/HighlightsMatch.java:
```diff
@@ -180,47 +180,16 @@ void addHit(String field, int startPos, int endPos, int startOffset, int endOffs
+   *
+   * @param startPosition The start position
+   * @param startOffset The start offset
+   * @param endPosition The end positions
```
```suggestion
   * @param endPosition The end position
```

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java:
```diff
@@ -163,11 +156,6 @@ public boolean equals(Object o) {
       FieldAndWeight that = (FieldAndWeight) o;
       return Float.compare(that.weight, weight) == 0 && Objects.equals(field, that.field);
     }
```
Did you note remove this equals override because the record would behave differently? (I tried to quickly look up what record does wrt equality for floats and doubles but couldn't find it, will look later)

mikemccand replied:
> (I tried to quickly look up what record does wrt equality for floats and doubles but couldn't find it, will look later)

I found one mention of floating point numbers in the [JEP for records](https://openjdk.org/jeps/395):

```
In addition, for all record classes the implicitly declared equals method is implemented so that it is reflexive and that it behaves consistently with hashCode for record classes that have floating point components. Again, explicitly declared equals and hashCode methods should behave similarly.
```

I'm not sure I understand that paragraph :)

uschindler replied:
Hi,
The equals methods for floats is generated by invokedynamic using a helper. When a record component is a float, the autogenerated impl is generated at runtime as combination of several Method handles. It compares them using this code:

https://github.com/openjdk/jdk/blob/9785e19f3f87306cabc26a862d35b89d41cfef93/src/java.base/share/classes/java/lang/runtime/ObjectMethods.java#L157

This is equivalent to the already existing code. So please remove all custom equals methods.

Unfortunately in the merged PR there are so one more. I will add a not on them. Please remove them in a followup.

shubhamvishu replied:
Nice! I'll check and followup with removing others I could find.

jpountz commented on lucene/test-framework/src/java/org/apache/lucene/tests/codecs/asserting/AssertingLiveDocsFormat.java:
```diff
@@ -68,9 +68,7 @@ public String toString() {
-  static class AssertingBits implements Bits {
-    final Bits in;
-
+  record AssertingBits(Bits in) implements Bits {
```
While it's not incorrect, I'm not sure if this class fits my mental model for a record as it's more about behaviors than about packaging other objects tohether.

shubhamvishu replied:
I agree, lets keep it as is.

jpountz commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/AssertingLeafReader.java:
```diff
@@ -1357,15 +1357,8 @@ public int getDocCount() {
-      this.in = in;
-    }
+  record AssertingPointTree(PointValues pointValues, PointValues.PointTree in)
+      implements PointValues.PointTree {
```
likewise: keep it a regular class?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/FixedBits.java:
```diff
@@ -17,15 +17,7 @@
-    this.bits = bits;
-    this.length = length;
-  }
+record FixedBits(long[] bits, int length) implements Bits {
```
it doesn't really fit my mental model for a record, maybe undo the refactoring on this class?

shubhamvishu replied:
Sure!

jpountz reviewed (approved):
I left some minor comments but the change looks good to me in general. Can you add an entry to lucene/MIGRATE.txt that explains that TopDocs, CollectionStatistics, TermStatistics and LeafMetadata were converted to records? I'm only considering mentioning these because these are the most user-facing APIs that are migrated to a record from what I can tell (please let me know if you think I'm missing an important one).

shubhamvishu committed:
Address comments

shubhamvishu committed:
Merge branch 'main' into use-record-classes

shubhamvishu committed:
Add MIGRATE and CHANGES entry

shubhamvishu commented:
@jpountz I addressed your comments and added a CHANGES, MIGRATE entry now. Thanks for reviewing!

mikemccand commented on lucene/CHANGES.txt:
```diff
@@ -118,6 +118,8 @@ API Changes
   argument with a `FacetsCollectorManager` and update the return type to include both `TopDocs` results as well as
   facets results. (Luca Cavanna)
 
+* GITHUB#13328: Convert CollectionStatistics, TermStatistics and LeafMetadata etc. to record classes. (Shubham Chaudhary)
```
Maybe reword to `Convert many basic Lucene classes to record classes, including X, Y, and Z`?  It's not just these three classes, if I'm reading the PR correctly, and the `etc.` should not have so much power ;)

shubhamvishu replied:
This looks better indeed. I'll reword it.

mikemccand commented on lucene/MIGRATE.md:
```diff
@@ -193,6 +193,7 @@ access the members using method calls instead of field accesses. Affected classe
 
 - `IOContext`, `MergeInfo`, and `FlushInfo` (GITHUB#13205)
 - `BooleanClause` (GITHUB#13261)
+- `CollectionStatistics`, `TermStatistics` and `LeafMetadata` (GITHUB#13328)
```
And  maybe change this wording too (same as `CHANGES.txt`)?  I think @jpountz asked for `TopDocs` to also be mentioned?

shubhamvishu replied:
Sure, I'll reword. I haven't changed `TopDocs` in this PR and hence I didn't include it. Maybe some confusion(prolly because there are just too many classes changed here)

mikemccand reviewed (approved):
Thank you for persisting here @shubhamvishu -- this is a massive rote refactoring change.  We should rather bias towards quickly merging such changes since they pick up conflicts so quickly.  I just left some minor feedback about wording in CHANGES and MIGRATE, else +1 to merge.

shubhamvishu committed:
Rephrase entries

shubhamvishu committed:
Merge branch 'main' into use-record-classes

stefanvodita commented:
I think we've run into a conflict with #13689. @jpountz - should I fix that or are you already doing it?

jpountz commented:
@javanna just fixed it, thanks @javanna !

uschindler commented on lucene/core/src/java/org/apache/lucene/search/MaxScoreAccumulator.java:
```diff
@@ -101,11 +93,6 @@ public boolean equals(Object o) {
       return docBase == result.docBase && Float.compare(result.score, score) == 0;
     }
 
-    @Override
```
Please also remove the equals, see this comment: https://github.com/apache/lucene/pull/13328#discussion_r1752842660

uschindler replied:
To string is also useless here. Please remove, too

shubhamvishu replied:
Sure, I'll fix it in a followup PR. I could only find this instance in the classes changed to records where we have removable custom equals/toString but I'll take another pass to be sure.

shubhamvishu replied:
I opened a followup PR #13769. Thanks!

uschindler reviewed (commented):
There is one as symmetry in removal of equals/hashcode.

Please fix this in a new PR.
